README.md
# Flappy Bird Clone 
Flappy Bird clone is a PC version of the original Flappy Bird game built for learning purposes (and fun!)  

This project is non-commercial. All the rights to Flappy Bird title and assets used in this project belong to the appropriate right holders.

## Built with 
* [Pygame](https://www.pygame.org/news)

## Author 
Ivan


CHANGELOG.md
# Changelog

## [0.7.0] - 2019-08-11
### Changed
- Image class constructor requires rect and rect_object.
- Refactor Numbers class.
- Refactor main logic.

## [0.6.0] - 2019-08-10
### Changed
- Refactor menus module.

### Removed
- Disclaimer message.

## [0.5.0] - 2019-07-24
### Added
- Initial version.

[0.7.0]: https://github.com/isylkin/flappy_bird_clone/compare/0.6.0...0.7.0
[0.6.0]: https://github.com/isylkin/flappy_bird_clone/compare/0.5.0...0.6.0
[0.5.0]: https://github.com/isylkin/flappy_bird_clone/compare/1576c33...0.5.0

lvl.py
import random
import pygame
import constants
from typing import Tuple, List
from assets_handler import Image, Numbers
from bird import Bird


class Level:
    def __init__(self, bird: Bird) -> None:
        self.bird = bird

        self.background_image = None
        self.background_scroll_speed = constants.BACKGROUND_SCROLL_SPEED
        self.background_x = constants.BACKGROUND_X

        self.pipe_sprites = pygame.sprite.Group()
        self.pipe_assets = constants.PIPE_ASSETS
        self.pipe_scroll_speed = constants.PIPE_SCROLL_SPEED
        self.increase_speed_milestone = constants.INCREASE_SPEED_MILESTONE

        self.score = 0
        self.nums = Numbers(constants.NUMBERS_ASSETS,
                            constants.BIG_NUMS_SPRITE_COORDS,
                            constants.BIG_NUMS_POSITION,
                            constants.SMALL_NUMS_SPRITE_COORDS,
                            constants.SMALL_NUMS_POSITION)

    def update(self) -> None:
        self._scroll_pipes()
        self._update_score()

    def _scroll_pipes(self) -> None:
        self._increase_pipe_scroll_speed(self.increase_speed_milestone)
        for pipe in self.pipe_sprites:
            pipe.rect[0] -= self.pipe_scroll_speed

    def _increase_pipe_scroll_speed(self, milestone: int) -> None:
        if self.score == milestone:
            self.pipe_scroll_speed = 3

    def _update_score(self) -> None:
        pipe_obstacles = self.pipe_sprites.sprites()[::2]  # One obstacle consists of two pipes
        for pipe in pipe_obstacles:
            if self._is_obstacle_passed(self.bird, obstacle=pipe):
                self.score += 1
                self.bird.sfx_point.play()
        self.score_sprites = self.nums.generate_score(self.score,
                                                      is_big=True)

    def _is_obstacle_passed(self, bird: Bird,
                            obstacle: pygame.sprite.Sprite) -> bool:
        if obstacle.rect.x == bird.rect.left - constants.PIPE_WIDTH:
            return True
        return False

    def draw(self, screen: pygame.Surface) -> None:
        self._scroll_background_image(screen)
        self.pipe_sprites.draw(screen)
        self.score_sprites.draw(screen)

    def _scroll_background_image(self, screen: pygame.Surface) -> None:
        # Create a rel_x value which stores a remainder of bkgd_x and bkgd_width
        # Blit the image on the pos rel_x - bkgd_width
        # Blit the same image on the rel_x position to scroll bkgd smoothly
        rel_x = self.background_x % self.background_image.get_rect().width
        screen.blit(self.background_image,
                    (rel_x - self.background_image.get_rect().width, 0))
        if rel_x < constants.SCREEN_WIDTH:
            screen.blit(self.background_image, (rel_x, 0))
        self.background_x -= self.background_scroll_speed


class Level01(Level):
    def __init__(self, bird: Bird, background: pygame.Surface) -> None:
        super().__init__(bird)
        self.background_image = background

        # List with pipes and their locations
        self.level = []
        self.level.extend(constants.INITIAL_PIPES)
        self.max_obstacle_height = constants.MAX_OBSTACLE_HEIGHT
        self.pipe_width = constants.PIPE_WIDTH
        self.pipes_break = constants.PIPES_BREAK

    def create_level(self) -> None:
        self._create_obstacles()
        self._create_pipe_sprites()

    def _create_obstacles(self) -> None:
        for i in range(1, 200):
            top_pipe, bottom_pipe = self._generate_pipes_coordinates()
            self.level.append(top_pipe)
            self.level.append(bottom_pipe)

    def _generate_pipes_coordinates(self) -> Tuple[List, List]:
        top_pipe_height, bottom_pipe_height = self._calculate_pipe_heights()
        pipes_x = self._get_pipes_x()

        top_pipe = [(0, abs(top_pipe_height - self.max_obstacle_height),
                     self.pipe_width, top_pipe_height), pipes_x, 0]
        bottom_pipe = [(self.pipe_width, 0, self.pipe_width, bottom_pipe_height),
                       pipes_x, abs(constants.SCREEN_HEIGHT - bottom_pipe_height)]
        return top_pipe, bottom_pipe

    def _calculate_pipe_heights(self) -> Tuple[int, int]:
        top_height = random.randint(80, self.max_obstacle_height)
        bottom_height = constants.SCREEN_HEIGHT - self.pipes_break - top_height
        return top_height, bottom_height

    def _get_pipes_x(self) -> int:
        last_pipe_right_border = self.level[len(self.level) - 1][1]
        pipes_x = last_pipe_right_border + self.pipe_width * 3
        return pipes_x

    def _create_pipe_sprites(self) -> None:
        for pipe in self.level:
            p = Image(self.pipe_assets,
                      sprite_coords=pipe[0],
                      rect=(pipe[1], pipe[2]))
            self.pipe_sprites.add(p)


bird.py
import pygame
import constants
from typing import Tuple, List
from assets_handler import SpriteSheet


class Bird(pygame.sprite.Sprite):
    def __init__(self, width: int, height: int, jump_speed: int,
                 starting_position: Tuple[int, int],
                 sfx_wing_path: str, sfx_point_path: str,
                 sfx_hit_path: str, sfx_die_path: str,
                 sprite_sheet_path: str, frames_coordinates: List) -> None:
        super().__init__()
        self.level = None

        self.width = width
        self.height = height
        self.jump_speed = jump_speed
        self.fall_speed = 0
        self.x, self.y = starting_position

        self.animation_frames = []
        sprite_sheet = SpriteSheet(sprite_sheet_path)
        for coord in frames_coordinates:
            frame = sprite_sheet.get_image(x=coord['x'], y=coord['y'],
                                           width=self.width,
                                           height=self.height)
            self.animation_frames.append(frame)

        self.sfx_wing = pygame.mixer.Sound(sfx_wing_path)
        self.sfx_point = pygame.mixer.Sound(sfx_point_path)
        self.sfx_hit = pygame.mixer.Sound(sfx_hit_path)
        self.sfx_die = pygame.mixer.Sound(sfx_die_path)

        self.is_obstacle_hit = False

    @property
    def rect(self) -> pygame.Rect:
        return pygame.Rect(self.x, self.y, self.width, self.height)

    @property
    def jump(self) -> None:
        if not self.is_obstacle_hit:
            self.sfx_wing.play()
            self.fall_speed = self.jump_speed

    def update(self, delta_time: float) -> None:
        # Increase falling speed with time
        self.y -= self.fall_speed * delta_time
        self.fall_speed -= 50 * delta_time

        if self.rect.top < 0:
            self.rect.top = 0
            self.fall_speed = 10  # Bounce from top
            self.fall_speed -= 50 * delta_time

        if self.rect.bottom >= constants.SCREEN_HEIGHT:
            self.fall_speed = 0
            self.rect.bottom = constants.SCREEN_HEIGHT

            self.sfx_hit.play()
            self.sfx_die.play()
            self.is_obstacle_hit = True
            return

        pipe_hit_list = pygame.sprite.spritecollide(self, self.level.pipe_sprites, False)
        for pipe in pipe_hit_list:
            self.level.pipe_scroll = 0
            self.level.bkg_scroll = 0
            self.animation_frames = [self.animation_frames[1]] * 3

            self.sfx_hit.play()
            self.sfx_die.play()
            self.is_obstacle_hit = True
            return

    @property
    def animate(self):
        game_time = pygame.time.get_ticks()
        if game_time % 400 >= 200:
            return self.animation_frames[1]
        elif game_time % 400 >= 100:
            return self.animation_frames[2]
        else:
            return self.animation_frames[0]


main.py
import time
import pygame
import menus
import constants
from lvl import Level01
from bird import Bird


def main_menu_stage(main_menu: menus.MainMenu,
                    screen: pygame.Surface) -> None:
    while not main_menu.is_started:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                raise SystemExit('QUIT')
            elif event.type == pygame.MOUSEBUTTONUP and main_menu.is_hovered:
                main_menu.is_started = True
        main_menu.update()
        main_menu.draw(screen)
        pygame.display.flip()


def get_ready_stage(get_ready_menu: menus.GetReadyMenu,
                    screen: pygame.Surface, bird: Bird) -> None:
    while not get_ready_menu.is_tapped:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                raise SystemExit('QUIT')
            elif event.type == pygame.KEYUP and event.key in (pygame.K_PAUSE, pygame.K_p):
                paused = not paused
            elif event.type == pygame.MOUSEBUTTONUP or (event.type == pygame.KEYUP and
                                                        event.key in (pygame.K_UP,
                                                        pygame.K_RETURN, pygame.K_SPACE)):
                bird.jump
                get_ready_menu.is_tapped = True
        get_ready_menu.draw(screen)
        pygame.display.flip()


def game_over_stage(game_over_screen: menus.GameOverScreen,
                    screen: pygame.Surface, current_level: Level01) -> None:
    while not game_over_screen.restart:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                raise SystemExit('QUIT')
            elif event.type == pygame.MOUSEBUTTONUP and game_over_screen.is_hovered:
                game_over_screen.restart = True
        game_over_screen.update()
        game_over_screen.draw(screen, current_level.score)
        pygame.display.flip()


def main() -> None:
    pygame.init()
    pygame.font.init()
    DISPLAY = (constants.SCREEN_WIDTH, constants.SCREEN_HEIGHT)
    screen = pygame.display.set_mode(DISPLAY)
    background = constants.BACKGROUND.convert()
    bird = Bird(constants.BIRD_WIDTH, constants.BIRD_HEIGHT,
                constants.BIRD_JUMP_SPEED,
                constants.BIRD_STARTING_POSITION,
                constants.SFX_WING_PATH, constants.SFX_POINT_PATH,
                constants.SFX_HIT_PATH, constants.SFX_DIE_PATH,
                constants.BIRD_SPRITE_SHEET_PATH,
                constants.BIRD_FRAMES_COORDINATES)

    level_list = list()
    level_list.append(Level01(bird, background))
    current_level_no = 0
    current_level = level_list[current_level_no]
    current_level.create_level()
    bird.level = current_level

    clock = pygame.time.Clock()
    main_menu = menus.MainMenu(bird, background, constants.MENU_ASSETS)
    get_ready_menu = menus.GetReadyMenu(bird, background, constants.MENU_ASSETS)
    game_over_screen = menus.GameOverScreen()

    # Game start
    main_menu_stage(main_menu, screen)
    while True:
        get_ready_stage(get_ready_menu, screen, bird)
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                raise SystemExit('QUIT')
            elif event.type == pygame.KEYUP and event.key in (pygame.K_PAUSE, pygame.K_p):
                paused = not paused
            elif event.type == pygame.MOUSEBUTTONUP or (event.type == pygame.KEYUP and
                                                        event.key in (pygame.K_UP, pygame.K_RETURN, pygame.K_SPACE)):
                bird.jump
        start_time = time.time() - 4
        delta_time = start_time - time.time()

        current_level.update()
        current_level.draw(screen)

        screen.blit(bird.animate, bird.rect)
        bird.update(delta_time / 100)

        if bird.is_obstacle_hit:
            game_over_stage(game_over_screen, screen, current_level)
            break

        clock.tick(65)
        pygame.display.flip()


if __name__ == '__main__':
    while True:
        main()


menus.py
import pygame
import constants
from bird import Bird
from assets_handler import Image, Numbers


class BaseMenu:
    def __init__(self, bird: Bird, background: pygame.Surface,
                 menu_assets_path: str) -> None:
        self.bird = bird
        self.background_image = background
        self.menu_assets = menu_assets_path
        self.background_x = constants.BACKGROUND_X
        self.background_scroll_speed = constants.BACKGROUND_SCROLL_SPEED
        self.sprite_list = pygame.sprite.Group()

    def _scroll_background_image(self, screen: pygame.Surface) -> None:
        rel_x = self.background_x % self.background_image.get_rect().width
        screen.blit(self.background_image,
                    (rel_x - self.background_image.get_rect().width, 0))
        if rel_x < constants.SCREEN_WIDTH:
            screen.blit(self.background_image, (rel_x, 0))
        self.background_x -= self.background_scroll_speed

    def draw(self, screen: pygame.Surface):
        self._scroll_background_image(screen)
        self.sprite_list.draw(screen)
        screen.blit(self.bird.animate, self.bird.rect)


class MainMenu(BaseMenu):
    def __init__(self, bird: Bird, background: pygame.Surface,
                 menu_assets_path: str) -> None:
        super().__init__(bird, background, menu_assets_path)
        self.start_button = Image(self.menu_assets,
                                  constants.START_BUTTON_SPRITE_COORDS,
                                  constants.START_BUTTON_POSITION,
                                  constants.START_BUTTON_COORDINATES)
        self.logo = Image(self.menu_assets,
                          constants.LOGO_SPRITE_COORDS,
                          constants.LOGO_POSITION)

        self.sprite_list.add(self.start_button)
        self.sprite_list.add(self.logo)

        self.is_hovered = False
        self.is_started = False

    def update(self):
        # Change hover status and cursor on hover over a start button
        if self.start_button.rect_object.collidepoint(pygame.mouse.get_pos()):
            pygame.mouse.set_cursor(*pygame.cursors.tri_left)
            self.is_hovered = True
        else:
            pygame.mouse.set_cursor(*pygame.cursors.arrow)
            self.is_hovered = False


class GetReadyMenu(BaseMenu):
    def __init__(self, bird: Bird, background: pygame.Surface,
                 menu_assets_path: str) -> None:
        super().__init__(bird, background, menu_assets_path)
        self.get_ready_button = Image(self.menu_assets,
                                      constants.GET_READY_BUTTON_SPRITE_COORDS,
                                      constants.GET_READY_BUTTON_POSITION)
        self.tap_tip = Image(self.menu_assets,
                             constants.TAP_TIP_SPRITE_COORDS,
                             constants.TAP_TIP_POSITION)

        self.sprite_list.add(self.get_ready_button)
        self.sprite_list.add(self.tap_tip)

        self.is_tapped = False


class GameOverScreen:
    def __init__(self) -> None:
        self.sprite_list = pygame.sprite.Group()
        self.menu_assets = constants.MENU_ASSETS
        self.game_over_image = Image(self.menu_assets,
                                     constants.GAME_OVER_SPRITE_COORDS,
                                     constants.GAME_OVER_POSITION)
        self.summary_banner = Image(self.menu_assets,
                                    constants.SUMMARY_SPRITE_COORD,
                                    constants.SUMMARY_POSITION)
        self.ok_button = Image(self.menu_assets,
                               constants.OK_BUTTON_SPRITE_COORDS,
                               constants.OK_BUTTON_POSITION,
                               constants.OK_BUTTON_COORDINATES)

        self.sprite_list.add(self.game_over_image)
        self.sprite_list.add(self.summary_banner)
        self.sprite_list.add(self.ok_button)

        self.is_hovered = False
        self.restart = False

        self.nums = Numbers(constants.NUMBERS_ASSETS,
                            constants.BIG_NUMS_SPRITE_COORDS,
                            constants.BIG_NUMS_POSITION,
                            constants.SMALL_NUMS_SPRITE_COORDS,
                            constants.SMALL_NUMS_POSITION)

    def draw(self, screen: pygame.Surface, score: int):
        self.sprite_list.draw(screen)
        score_sprites = self._generate_score_sprites(score)
        score_sprites.draw(screen)

    def _generate_score_sprites(self, score):
        score_sprites = self.nums.generate_score(score, is_big=False)
        return score_sprites

    def update(self):
        if self.ok_button.rect_object.collidepoint(pygame.mouse.get_pos()):
            pygame.mouse.set_cursor(*pygame.cursors.tri_left)
            self.is_hovered = True
        else:
            pygame.mouse.set_cursor(*pygame.cursors.arrow)
            self.is_hovered = False


get_sprite.py
"""
Module for pulling sprites from sprite sheets
of required width and height.
"""

import pygame
import constants


class SpriteSheet(object):
    def __init__(self, file_name):

        # Load the sprite sheet
        self.sprite_sheet = pygame.image.load(file_name).convert()

    def get_image(self, x, y, width, height):
        """ Grab a single image out of a larger spritesheet
            Pass in the x, y location of the sprite
            and the width and height of the sprite. """

        # Create a new blank image
        image = pygame.Surface([width, height]).convert()

        # Copy the sprite from large img to the smaller one
        image.blit(self.sprite_sheet, (0, 0), (x, y, width, height))

        # Black as the transparent color
        image.set_colorkey(constants.VIOLET)

        return image


assets_handler.py
import pygame
import constants
from typing import Tuple, Dict, List


class SpriteSheet:
    def __init__(self, file_name: str) -> None:
        self.sprite_sheet = pygame.image.load(file_name).convert()

    def get_image(self, x: int, y: int, width: int, height: int) -> pygame.Surface:
        image = pygame.Surface([width, height]).convert()
        image.blit(source=self.sprite_sheet,
                   dest=(0, 0), area=(x, y, width, height))
        image.set_colorkey(constants.VIOLET)  # Violet as the transparent color
        return image


class Image(pygame.sprite.Sprite):
    def __init__(self, assets_path: str,
                 sprite_coords: Tuple[int, int, int, int],
                 rect: Tuple[int, int],
                 rect_object: pygame.Rect = None) -> None:
        super().__init__()
        sprite_sheet = SpriteSheet(assets_path)
        self.image = sprite_sheet.get_image(sprite_coords[0],
                                            sprite_coords[1],
                                            sprite_coords[2],
                                            sprite_coords[3],)
        self._rect = self.image.get_rect()
        self._rect.x = rect[0]
        self._rect.y = rect[1]

        self._rect_object = rect_object

    @property
    def rect(self):
        return self._rect

    @rect.setter
    def rect(self, value):
        self._rect = self.image.get_rect()
        self._rect.x = value[0]
        self._rect.y = value[1]

    @property
    def rect_object(self):
        return self._rect_object

    @rect_object.setter
    def rect_object(self, value):
        self._rect_object = value


class Numbers:
    def __init__(self, nums_assets_path: str,
                 big_nums_coords: List, big_nums_position: Tuple [int, int],
                 small_nums_coords: List, small_nums_position: Tuple[int, int]) -> None:
        self.nums_assets = nums_assets_path
        self.big_nums_coords = big_nums_coords  # Live score
        self.big_nums_position = big_nums_position
        self.small_nums_coords = small_nums_coords  # Summary
        self.small_nums_position = small_nums_position
        self.big_nums_images = {}
        self.small_nums_images = {}

        for num in range(len(self.big_nums_coords)):
            self.big_nums_images[f'{num}'] = Image(self.nums_assets,
                                                   self.big_nums_coords[num],
                                                   self.big_nums_position)
            self.small_nums_images[f'{num}'] = Image(self.nums_assets,
                                                     self.small_nums_coords[num],
                                                     self.small_nums_position)

    def generate_score(self, score: int, is_big: bool) -> pygame.sprite.Group:
        score_sprites = pygame.sprite.Group()
        str_score = str(score)
        nums_images, nums_coords, nums_position = self._get_nums_data(is_big)

        first_number = nums_images[str_score[0]]
        pos_x = first_number.rect[0]
        score_sprites.add(first_number)

        self._add_extra_numbers(pos_x, str_score, nums_images,
                                nums_coords, nums_position, score_sprites)

        return score_sprites

    def _get_nums_data(self, is_big: bool) -> Tuple:
        if is_big:
            nums_images = self.big_nums_images
            nums_coords = self.big_nums_coords
            nums_position = self.big_nums_position
        else:
            nums_images = self.small_nums_images
            nums_coords = self.small_nums_coords
            nums_position = self.small_nums_position
        return nums_images, nums_coords, nums_position

    def _add_extra_numbers(self, pos_x: int, score: str, nums_images: Dict,
                           nums_coords: List, nums_position: Tuple,
                           score_sprites: pygame.sprite.Group) -> None:
        if len(score) > 1:
            for i in range(1, len(score)):
                if score[i] == score[0] or score[i] == score[i - 1]:
                    extra_number = Image(self.nums_assets,
                                         nums_coords[i],
                                         nums_position)
                else:
                    extra_number = nums_images[score[i]]
                pos_x += extra_number.rect.width + 2
                extra_number.rect = (pos_x, nums_position[1])
                score_sprites.add(extra_number)


constants.py
import os
import pygame

FPS = 60
SCREEN_WIDTH = 854
SCREEN_HEIGHT = 480

# COLORS
BLACK = (0, 0, 0)
VIOLET = (128, 0, 128)

# PIPES
MAX_OBSTACLE_HEIGHT = 260
PIPE_WIDTH = 100
PIPES_BREAK = 140
PIPE_SCROLL_SPEED = 2
INCREASE_SPEED_MILESTONE = 20
PIPE_COLOR = '#004400'
PIPE_ASSETS = os.path.join("assets", "pipe.png")
INITIAL_PIPES = [[(0, 80, 100, 180), 900, 0],
                [(100, 0, 100, 160), 900, 320],]

# BIRD
MAIN_MENU_POSITION = (395, 235)
BIRD_STARTING_POSITION = (388, 231)
BIRD_WIDTH = 57
BIRD_HEIGHT = 39
BIRD_JUMP_SPEED = -70
SFX_WING_PATH = os.path.join("assets", "sfx", "sfx_wing.wav")
SFX_POINT_PATH = os.path.join("assets", "sfx", "sfx_point.wav")
SFX_HIT_PATH = os.path.join("assets", "sfx", "sfx_hit.wav")
SFX_DIE_PATH = os.path.join("assets", "sfx", "sfx_die.wav")
BIRD_SPRITE_SHEET_PATH = os.path.join("assets", "bird.png")
BIRD_FRAMES_COORDINATES = [{'x': 0, 'y': 0},
                           {'x': 97, 'y': 0},
                           {'x': 193, 'y': 0},]

# BACKGROUND
BACKGROUND = pygame.image.load(os.path.join("assets", "bg.png"))
BACKGROUND_SCROLL_SPEED = 0.5
BACKGROUND_X = 0

# MENUS
MENU_ASSETS = os.path.join("assets", "menu.png")
START_BUTTON_SPRITE_COORDS = (345, 763, 144, 52)
START_BUTTON_POSITION = (345, 345)
START_BUTTON_COORDINATES = pygame.Rect(345, 345, 144, 52)

LOGO_SPRITE_COORDS = (0, 955, 500, 144)
LOGO_POSITION = (177, 50)

GET_READY_BUTTON_SPRITE_COORDS = (0, 792, 312, 85)
GET_READY_BUTTON_POSITION = (275, 40)

TAP_TIP_SPRITE_COORDS = (102, 430, 145, 184)
TAP_TIP_POSITION = (460, 153)

GAME_OVER_SPRITE_COORDS = (0, 710, 340, 75)
GAME_OVER_POSITION = (270, 8)

SUMMARY_SPRITE_COORD = (0, 202, 411, 215)
SUMMARY_POSITION = (230, 140)

OK_BUTTON_SPRITE_COORDS = (357, 477, 150, 54)
OK_BUTTON_POSITION = (357, 390)
OK_BUTTON_COORDINATES = pygame.Rect(357, 390, 150, 54)

# NUMBERS
NUMBERS_ASSETS = os.path.join("assets", "nums.png")
BIG_NUMS_SPRITE_COORDS = [(4, 61, 24, 33), (7, 122, 24, 33), (7, 178, 24, 33),
                          (7, 233, 24, 33), (0, 312, 24, 33), (0, 353, 24, 33),
                          (2, 0, 24, 33), (36, 0, 24, 33), (71, 0, 24, 33),
                          (105, 0, 24, 33)]
BIG_NUMS_POSITION = (SCREEN_WIDTH / 2 - 10, 10)

SMALL_NUMS_SPRITE_COORDS = [(356, 0, 22, 25), (2, 273, 22, 24), (137, 0, 22, 25),
                            (164, 0, 22, 25), (191, 0, 22, 25), (219, 0, 22, 25),
                            (247, 0, 22, 25), (274, 0, 22, 25), (302, 0, 22, 25),
                            (329, 0, 22, 25)]
SMALL_NUMS_POSITION = (553, 205)